// gamemodes/Default.js
var PLAYER_SPRITES = [
	"neko_eiji"
	/*"neko",
	"neko_big",
	"neko_eiji"
	"car_test",
	"little",
	"bloc",
	"player",
	"wisp",
	"onclick",
	"mini_cube"*/
];

var Default = function() {
	this.params = {
		"player": {
			"speed": 0.9,
			"acc": 0.07,
			"jumpPower": 0.5,
			"maxSpeedLimiter": "strict",
			"deceleration": 0.001,
			"actions": {
				"UP": "jump",
				"LEFT":"move",
				"DOWN": "",
				"RIGHT": "move",
				"ACTION1": "jump",
				"ACTION2": "jump"
			}
		}
		/*
		maxSpeedLimiter
			Change the way the velocity is limited when the speed has reach its maximum

			strict : Don't modify the velocity if the maximum speed is already reached (used when there's a gravity)
			normal : The angle is modified but the overall speed is not increased (may cause the player to plan instead of falling hen there's a gravity)
		*/
	};
};




Default.prototype.onPlayerJoin = function(player) {
	// body...
};

Default.prototype.onPlayerLeave = function(player) {
	// body...
};




Default.prototype.getPlayerEntity = function(player) {
	return PLAYER_SPRITES[Math.floor(Math.random()*PLAYER_SPRITES.length)];
};




Default.prototype.playerActionStart = function(action, player) {
	switch (this.params.player.actions[action]) {
		case "move":
			this.moveEntity(action, player.entity, true);
			break;
		case "jump":
			this.jumpEntity(player.entity);
			break;
	}
};

Default.prototype.playerActionStop = function(action, player) {
	switch (this.params.player.actions[action]) {
		case "move":
			this.moveEntity(action, player.entity, false);
			break;
	}
};


Default.prototype.keyDown = function(key, player) {
	// body...
};

Default.prototype.keyUp = function(key, player) {
	// body...
};




Default.prototype.moveEntity = function(direction, entity, enable) {
	enable = Boolean(enable);
	direction = direction.toLowerCase();

	entity.move.direction[direction] = enable;

	var moving = false;
	for (var id in entity.move.direction) {
		if (entity.move.direction[id]) {
			moving = true;
			break;
		}
	}
	if (moving) {
		entity.move.speed = this.params.player.speed;
		entity.move.acc = this.params.player.acc;
	}
	else {
		entity.move.speed = 0;
		entity.move.acc = 0;
	}
};

Default.prototype.jumpEntity = function(entity) {
	entity.jump = this.params.player.jumpPower;
};






module.exports = Default;
